### ISSUE TITLE
Change: externalize TNG hooks-context

### ISSUE THREAD

**Issue by getify** [getify/TNG-Hooks#20]
Going along with some of the ideas proposed around hooks being designed as "algebraic effects", I am contemplating the idea of changing TNG hooks-context to work externally to the Articulated Function rather than being saved internally. This would seemingly make hooks both more compatible with FP, but also make them more testable and debuggable.

I'm opening this issue to start sketching out ideas for what that might look like, and take feedback/debate on if that's a useful direction for TNG to go.

Here's a basic example as it currently works:

```js
function foo(val = 3) {
   var [x,updateX] = useState(val);
   updateX(v => v + 1);
   useEffect(function(){
      console.log(`x: ${x + 1}`);
   });
}

foo = TNG(foo);

foo();     // x: 4
foo();     // x: 5
foo();     // x: 6
```

So here's what I'm now considering instead. If you call an Articulated Function directly (aka, with no context), and don't apply its result context, it runs but does nothing (in terms of observable effects, etc):

```js
function foo(val = 3) {
   var [x,updateX] = useState(val);
   updateX(v => v + 1);
   useEffect(function(){
      console.log(`x: ${x + 1}`);
   });
}

foo = TNG(foo);

foo();    // ..nothing..
foo();    // ..nothing..
foo();    // ..nothing..
```

Each time you invoke an Articulated Function, it returns its own modified TNG hooks context. If you `apply()` that context, it will process any of its pending effects:

```js
var context = foo();
context.apply();  // x: 4
context = foo();
context.apply();  // x: 4
context = foo();
context.apply();  // x: 4
```

But notice that each direct invocation of an Articulated Function without a context object in the first argument position will then start with a brand new context (thus printing `"x: 4"` each time). If you provide a context object in the first argument position, it will use that context as its initial state:

```js
context = foo().apply();    // x: 4
context = foo(context).apply();    // x: 5
context = foo(context).apply();    // x: 6
```
 
**Note:** `apply()` returns the context as well, to make chaining of these method calls more ergonomic.

The biggest impact of this change will be that, to avoid intruding on the function signature (parameter list), we're hijacking the `this` context of Articulated Functions to communicate the TNG hooks-context, meaning Articulated Functions can't use their own `this` context.

That shouldn't be too much of a limitation, as TNG is really designed to be used on stand-alone functions, not `this`-aware methods, anyway.

We also lose the ability to have Articulated Functions return values. We probably could restore that by putting the return value on the context, like `foo().returned` or something. Not sure if that matters enough to be worth it?

It's also a bit more inconvenient to have to pass the context and call `.apply()` after, just to keep an Articulated Function stateful. So, for convenience, we can produce a context-wrapped version of an Articulated Function, so that it works as it always did originally:

```js
foo = foo.wrap();
foo();   // x: 4
foo();   // x: 5
foo();   // x: 6
```

This would be discouraged (given the implications it has on testability and debugability that motivated this whole change), but still provided for convenience and legacy reasons. And just for illustrative purposes, `foo.wrap()` is basically a simple helper like:

```js
foo.wrap = (function scope(){
   var context;
   return function wrapped(...args){
      context = foo(context,...args).apply();
   };
})();
```

----

Also, consider this more sophisticated example:

```js
function hitCounter(btnID) {
   // accessing the new current TNG hooks-context
   var context = this;

   var [count,updateCount] = useState(0);
   var elem = useRef();

   useEffect(function onInit(){
      elem.current = document.getElementById(btnID);
      elem.current.addEventListener("click",function onClick(){
         updateCount(x => x + 1);

         // re-render the button
         hitCounter(context,btnID).apply();
      },false);
   },[btnID]);

   useEffect(function onRender(){
      elem.value = count;
   },[count]);
}

hitCounter = TNG(hitCounter);

hitCounter("the-btn").apply();   // button initially says: "0"

// click the button, it now says: "1"
// cilck the button again, it now says: "2"
// cilck the button yet again, it now says: "3"
```

**Note:** an Articulated Function now has its `this` bound to its own new current TNG hooks context (**not any previous context** that would have been used to invoke the Articulated Function), for internal access, in this case for the re-render that the click handler does later; the context object `this` points at will be the same object that's returned from the current invocation of the Articulated Function.

**Comment by JesterXL** [comment]
- Having `foo` become a noop and not return a value feels super impure, implying it has side effects each time you call it.
- I like the `apply` concept to actually execute it when you're ready.
- ... however, it leads to confusion for old people like me who say "wait, `Function.apply` or a different apply... or is that how the magic works?"
- is there some place in a closure you can store state? `this` is negated if you switch to Arrow functions and you don't need any worries of `this`. People see `this` and think it is ok to use, and in the FP world, it's not.

**Comment by getify** [comment]
> Having foo become a noop and not return a value

`foo(..)` does return a value, it just happens to be the new/modified TNG context that it returns. I don't know of any other way to get that context "out" of the function other than to use the return value to do so.

Articulated Functions don't really "compute" things the way typical functions do. They're more like a "reducer" computing a new state, which is conceptually like the new context being returned.

> ... however, it leads to confusion for old people like me who

I did worry a bit about that. Since it's `apply()` on an object and not a function, was hoping that would help. Should it be renamed to like `affect()` or `run()` something?

> People see this and think it is ok to use, and in the FP world, it's not.

I don't like the use of `this` here, but I'm experimenting with it as a way to transport the context into your function, as opposed to having to intrude on the parameter list to do so.

**Comment by JesterXL** [comment]
Naw, this part:
```javascript
foo = TNG(foo);

foo();    // ..nothing..
foo();    // ..nothing..
foo();    // ..nothing..
```
That should be a ... thing... I don't know what, but if it's like an Object or Context or whatever that has your apply method, that's great.

`run` is cool... and a good idea. That's what it does; it runs just like a Promise does and runs and hides all the effects. You start using things like `ap` or `then` or `map` you start getting all the Category Theory peeps up in arms, then they start using math words at you. It's how Folktale runs tasks (cancelable Promises):

```javascript
const { task } = require('folktale/concurrency/task');

const delay = (ms) => task(
  (resolver) => {
    const timerId = setTimeout(() => resolver.resolve(ms), ms);
    resolver.cleanup(() => {
      clearTimeout(timerId);
    });
  }
);

// waits 100ms
const result = await delay(100).or(delay(2000)).run().promise();
$ASSERT(result == 100);
```

Good luck on the `this` thing, man, that's... hard, heh!

**Comment by getify** [comment]
> That should be a ... thing... I don't know what, but if it's like an Object or Context or whatever that has your apply method, that's great.

Sorry for my miscommunication, lemme clarify... those calls all would be returning the context object, exactly as in the next snippet.

The "nothing" there was intended to indicate that the console statements didn't run and print the "x: 4" type messages because the effects didn't run since you didn't call `apply()`. I'll figure out a way to make that clearer. Sorry!

**Comment by getify** [comment]
> `run` is cool... and a good idea. That's what it does

Since what it actually does is invoke the pending effects, I'm currently thinking I'll call it `effects()` instead of `apply()` or `run()`. I don't want there to be a connotation that the Articulated Function hasn't run yet -- it has, its effects are what are still pending.

**Comment by JesterXL** [comment]
Oh it's not lazy? Cool, yes, `effects` much better then. I'd prefer a verb, but that's `affect` which is a `canOfWorms` lol

**Comment by getify** [comment]
Yeah. My guess is, what FP'ers mostly don't like is eager side-effects, nor do they like eager long-running tasks. That's why I'm hoping that containing things like asynchrony to lazy effects, and memoized functions, will be more "FP friendly". :)

**Comment by pkoch** [comment]
Ok, this might sound dumb, but why don't we add another attachment to the returned `tngf` function (like `reset`) that gets you `buckets[tngf]`?

Doesn't get you the `effects` affordance, but your reasoning makes it feel like a consequence, not the target of the change.

**Comment by getify** [comment]
**Update:** work has resumed on this after a long hiatus. And things are looking good.

I think the main code is in good shape now. I think what's left is to re-work all the tests (and add many more, for coverage sake), and docs updates. That's going to be a lot of tedious work, but it should be relatively straightforward.
